File: library/world_canvas/src/world_canvas_libs/world_canvas_client_examples/src/test_annotation_collection.cpp
/*
 * test_annotation_collection.cpp
 *
 *  Created on: Jul 8, 2014
 *      Author: jorge
 */

#include <ros/ros.h>

#include <yocs_msgs/Wall.h>
#include <yocs_msgs/Column.h>
#include <nav_msgs/OccupancyGrid.h>

#include <world_canvas_client_cpp/annotation_collection.hpp>


int main(int argc, char** argv)
{
  ros::init(argc, argv, "test_annotation_collection");

  // Parameters
  ros::NodeHandle nh("~");

  std::string world_name;
  std::string topic_name;
  std::string topic_type;
  std::string default_tn("annotations");
  std::string default_tt;
  std::string default_wn("INVALID_WORLD");
  bool        pub_as_list;
  std::vector<std::string> uuids;
  std::vector<std::string> names;
  std::vector<std::string> types;
  std::vector<std::string> keywords;
  std::vector<std::string> relationships;

  nh.param("world",         world_name, default_wn);
  nh.param("topic_name",    topic_name, default_tn);
  nh.param("topic_type",    topic_type, default_tt);
  nh.param("pub_as_list",   pub_as_list, false);
  nh.param("uuids",         uuids, uuids);
  nh.param("names",         names, names);
  nh.param("types",         types, types);
  nh.param("keywords",      keywords, keywords);
  nh.param("relationships", relationships, relationships);

  // Prepare the annotation collection
  wcf::FilterCriteria filter(world_name, uuids, names, types, keywords, relationships);
  wcf::AnnotationCollection ac(filter);
  ac.loadData();
  ROS_INFO("Annotation collection ready!");

  // Publish annotations' visual markers on client side
  ac.publishMarkers("annotation_markers");

  // Request server to publish the annotations
  ac.publish(topic_name, topic_type, true, pub_as_list);

  std::vector<yocs_msgs::Wall> walls;
  std::vector<yocs_msgs::Column> columns;
  std::vector<nav_msgs::OccupancyGrid> maps;

  ROS_INFO("Done!  got %u walls, %u columns and %u maps",
           ac.getData(walls), ac.getData(columns), ac.getData(maps));
  ROS_INFO("(for confirmation    %lu   %lu   %lu)", walls.size(), columns.size(), maps.size());
  ros::Publisher wp = nh.advertise<yocs_msgs::Wall> ("walls_on_client", 1, true);
  ros::Publisher cp = nh.advertise<yocs_msgs::Column> ("columns_on_client", 1, true);
  ros::Publisher mp = nh.advertise<nav_msgs::OccupancyGrid> ("maps_on_client", 1, true);

  for (unsigned int i = 0; i < walls.size(); i++)
    wp.publish(walls[i]);
  for (unsigned int i = 0; i < columns.size(); i++)
    cp.publish(columns[i]);
  for (unsigned int i = 0; i < maps.size(); i++)
    mp.publish(maps[i]);

  ros::spin();

  return 0;
}

File: library/world_canvas/src/world_canvas_libs/world_canvas_client_examples/src/test_annotation_collection.py
#!/usr/bin/env python

import rospy
import copy

import world_canvas_client

from yocs_msgs.msg import *
from world_canvas_msgs.msg import *

if __name__ == '__main__':
    rospy.init_node('test_annotation_collection')
    topic_name  = rospy.get_param('~topic_name', 'annotations')
    topic_type  = rospy.get_param('~topic_type', None)
    pub_as_list = rospy.get_param('~pub_as_list', False)
    namespace   = rospy.get_param('~srv_namespace', '')
    world    = rospy.get_param('~world')
    uuids    = rospy.get_param('~uuids', [])
    names    = rospy.get_param('~names', [])
    types    = rospy.get_param('~types', [])
    keywords = rospy.get_param('~keywords', [])
    related  = rospy.get_param('~relationships', [])

    ac = world_canvas_client.AnnotationCollection(world, uuids, names, types, keywords, related, namespace)
    ac.load_data()
    walls = ac.get_data_of_type(yocs_msgs.msg.Wall)
    columns = ac.get_data_of_type(yocs_msgs.msg.Column)

    # Publish annotations' visual markers on client side
    ac.publish_markers('annotation_markers')

    # Publish annotations on client side
    ac.publish(topic_name + '_client', topic_type, False, pub_as_list)

    # Request server to also publish the same annotations
    ac.publish(topic_name,             topic_type, True,  pub_as_list)
    
    rospy.loginfo("Done")
    rospy.spin()
